fibers/timers.scm

```diff
@@ -45,23 +46,37 @@
   "Make an operation that will succeed when the current time is
 greater than or equal to @var{expiry}, expressed in internal time
 units.  The operation will succeed with no values."
-  (make-base-operation #f
-                       (lambda ()
+  (define wheel-entry
+    ;; If true, this is the currently active timer entry for this operation.
+    #f)
+
```

Comment:
This looks like a problem. Previously, operation objects are reusable, and nothing suggested you need to re-create the operation object (convenient for loops!). Now they aren't, and worse it's undocumented.

If state is needed, it needs to be moved into arguments of one of the 'lambdas' below (is API change, but you could define separate 'make-base-operation' and 'make-base-operation/stateful'). Another option is to rename 'flag' to state', make it a pair of (atomic box with flag . wheel-entry) and allow overriding the default construction of the state (`(fibers operations)` initialises to an atomic box, but it doesn't actually _use_ its contents -- doing something with it is left entirely to the individual implementations).

Reply:
I agree that passing state around would be more elegant.  In practice though, the current approach is okay IMO because the variable is closed over by the closures of the operation.

In this case, I fixed the problem you mentioned (being able to reuse a timer operation after it's been "canceled") simply by resetting the `timer-wheel` variable upon cancellation.

Reply:
> In practice though, the current approach is okay IMO because the variable is closed over by the closures of the operation.
> 

AFAICT, there is still a problem if a single timer operation value is used concurrently from multiple fibers. (Having `guard-operation` would make this pattern work.)

Reply:
> AFAICT, there is still a problem if a single timer operation value is used concurrently from multiple fibers. (Having guard-operation would make this pattern work.)

Apologies @LiberalArtist, I missed this comment of yours.

Would using an atomic box (as I suggested in the first message) solve the problem?

Reply:
I suppose it might be possible to adjust the timer wheel implementation to use atomic boxes everywhere + compare and swap, but that seems inefficient, and because _two_ pointers need to be replaced, difficult to do correctly and verify for correctness. It seems simpler & less error-prone to me to simply add a state argument.

In case of 'replace #f' by 'atomic box containing #f', no. We need to assign things to the right thread (in particular, the right scheduler, because of work stealing(?)) and atomic boxes don't do such things, they impose ordering constraints.

Reply:
(Note: IIUC, if/when `guard` is created, an explicit 'state' argument could be eliminated, although a 'state' operation-construction API could still be provided for convenience.)

Reply:
Oh, re atomicity, I was thinking about the `(set! timer-wheel ...)` bit, but you're saying the timer wheel implementation itself should be made atomic?

I must say I'm unclear on that, though my understanding is that there's one timer wheel per scheduler and one scheduler per thread, no?

@wingo?

Reply:
I'm not saying that it should be made atomic. The state version is also an option.

>though my understanding is that there's one timer wheel per scheduler and one scheduler per thread, no?

And that's why, if you don't go for the state version, it needs to be atomic -- you are saving a timer wheel entry of the current scheduler in the closure, and due to work stealing the fiber might migrate to another scheduler (IIUC), so the cancellation can be run from another scheduler, and fiddling with another threads data structures leads to trouble unless designed for that.